### DISCUSSION TITLE
Is there a way to convert response structs to JSON automatically?

### DISCUSSION THREAD

**Discussion by bubelov** [rwf2/Rocket#1810] [Questions] [answered]
I'm writing JSON API and I wonder if it's possible to simply return `Result<MyStruct, MyErrorStruct>` instead of `Result<Json<MyStruct>, MyErrorStruct>`.

`MyErrorStruct` implements custom `Responder`, so Rocket has no problem with the fact that it lacks `Json<>`. Implementing custom responders for all non-error structs doesn't seem like a good and scalable solution, but that's essentially what compiler expects and complains about. Any suggestions?

**Comment by jebrosen** [comment, marked as answer]
> I'm writing JSON API and I wonder if it's possible to simply return `Result<MyStruct, MyErrorStruct>` instead of `Result<Json<MyStruct>, MyErrorStruct>`.

Probably not. How would this approach work for applications that don't want to return JSON?

---

The recommended approach to repetitive response shapes such as yours is to implement [or derive](https://api.rocket.rs/v0.5-rc/rocket/derive.Responder.html) `Responder` on one of your own types, for instance:

```rust
enum ApiResult<T> {
    Ok(Json<T>),
    Redirect(Redirect),
    Err(MyErrorStruct),
}

// previously: -> Result<Json<MyStruct>, MyErrorStruct>.
fn route() -> ApiResult<MyStruct> { }
```

**Reply by bubelov** [reply]
> Probably not. How would this approach work for applications that don't want to return JSON?

Those structs derive `Serialize`, if that's not enough of a hint, there could be other ways to set the preferred content type globally. I worked with Spring Boot for a bit, and this thing just worked out of the box. It also used per-controller hints like this:

```kotlin
@RequestMapping(produces = ["application/json"])
```

Anyway, your approach worked like a charm, thanks for your help! I ended up with the following code:

```rust
#[derive(Responder)]
#[response(bound = "T: Serialize")]
pub enum ApiResult<T> {
    Ok(Json<T>),
    Err(Error),
}

impl<T> ApiResult<T> {
    pub fn new(result: Result<Option<T>, Report>) -> ApiResult<T> {
        match result {
            Ok(opt) => match opt {
                Some(val) => ApiResult::Ok(Json(val)),
                None => ApiResult::Err(Error::short(Status::NotFound)),
            },
            Err(e) => ApiResult::Err(Error::full(Status::InternalServerError, e)),
        }
    }
}
```

Which allowed me to cut the boilerplate in a route handler to:

```rust
#[get("/exchange_rates?<quote>&<base>")]
pub async fn get(quote: &str, base: &str, db: Db, _user: User) -> ApiResult<ExchangeRate> {
    ApiResult::new(exchange_rates::get_by_quote_and_base(quote, base, db).await)
}
```

That's almost as hassle-free as passing `Result<Option<T>, eyre::Report>` directly. Even if there is a way of converting `T` to JSON automatically, it probably won't work with a foreign struct such as `eyre::Report`, so it would need some boilerplate to convert foreign error struct to local struct which implements a `Responder` trait.

**Reply by jebrosen** [reply]
> Those structs derive Serialize, if that's not enough of a hint, there could be other ways to set the preferred content type globally. I worked with Spring Boot for a bit, and this thing just worked out of the box.

Ah, okay. Although it would be nice, I don't believe we can implement a similar approach in Rust - unlike in Java, which has extensive reflection capabilities at runtime. At compile time, "optionally" requiring a trait such as `Serialize` usually requires generics specialization, an unstable/nightly-only feature with several design problems to be resolved before it can be fully implemented and stabilized. Similarly at runtime, determining whether or not a type implements a certain trait requires metadata about the types that is not readily available. Rust does offer a few reflection-like APIs, but most of them are too limited, incomplete, or unstable to be used for this purpose in Rocket (at least, not yet).

**Reply by bubelov** [reply]
@jebrosen yea, Spring Boot just analyzes deps at runtime and makes opinionated decisions based on that, in a rather non-transparent manner, so its approach has its own downsides. I had a similar API implemented in Spring Boot and it took about 30 seconds to launch it on  a $5 Digital Ocean VPS. Just tried Rocket app today and it took a blink of an eye to launch. If that's the cost of being fast, it's fine by me =)

**Reply by JockeTF** [reply]
Sometimes you may also just want [type aliases](https://doc.rust-lang.org/book/ch19-04-advanced-types.html#creating-type-synonyms-with-type-aliases) to make things a bit more concise.

```rust
type ApiResult<T> = Result<Json<T>, Error>;
```

I'm not sure how relevant this is this case, but still!

**Reply by the10thWiz** [reply]
Actually, there is an interesting option here: add this as a feature of the responder derive. I image that would look something like this:
```
#[derive(Responder, Serialize)]
#[response(with = "Json")]
pub struct MyStruct {
    ...
}
```
The `with` option could be fairly generic, but the idea is that the responder implementation would look something like this:
```
impl Responder for MyStruct {
   fn respond_to(self, _) -> _ {
      Json(self).respond_to(_)
   }
}
```
The idea is that the `with` parameter is essentially a function (or wrapping type) that creates a struct that implements responder. Obviously, writing this implementation by hand isn't hard, but handling it via the derive macro like this could be convient.

**Comment by 3olkin** [comment]
You can also use `rocket::serde::json::Value` as response type, it also automatically add `content-type` header and in some cases can work exactly as you want. In addition `serde_json` has several [useful methods](https://docs.serde.rs/serde_json/value/index.html) for converting to `Value`.
Small example:
```rust
#[post("/login", format = "json", data = "<payload>")]
pub async fn login(
    payload: Json<LoginRequest>,
    secrets: &State<Secrets>,
    conn: PgConn,
) -> api::Result<Value> {
    let payload = payload.into_inner();
    let contractor = Contractor::find_by_email(payload.email, &conn).await?;
    if verify_password(&payload.password, &contractor.password) {
        let token = jwt::sign(contractor.uuid, "contractor", &secrets.jwt_secret_key)?;
        Ok(json!({
            "access_token": token,
            "token_type": "Bearer"
        }))
    } else {
        Err(Error::WrongCredentials.into())
    }
}
```

**Reply by 3olkin** [reply]
In addition you can convert any serializable struct to `serde_json::Value` this way: `json!(my_struct)`
